tls/s2n_ktls_io.c

```diff
@@ -0,0 +1,227 @@
+        struct cmsghdr _align;
+    } control_msg = { 0 };
+
+    msg.msg_control = control_msg.buf;
+    msg.msg_controllen = sizeof(control_msg.buf);
+#endif
+
+    RESULT_GUARD(s2n_ktls_set_ancillary_data(&msg, record_type));
+    RESULT_GUARD(s2n_ktls_send_msg_impl(sock, &msg, msg_iov, count, blocked, send_len));
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_recv_msg_impl(struct s2n_connection *conn, int sock, struct msghdr *msg,
+        struct iovec *msg_iov, s2n_blocked_status *blocked, ssize_t *bytes_read)
+{
+    RESULT_ENSURE_REF(msg);
+    RESULT_ENSURE_REF(msg_iov);
+    RESULT_ENSURE_REF(blocked);
+    RESULT_ENSURE_REF(bytes_read);
+
+    /* set receive buffer */
+    msg->msg_iov = msg_iov;
+    msg->msg_iovlen = 1;
+
+    *blocked = S2N_BLOCKED_ON_READ;
+    *bytes_read = recvmsg(sock, msg, 0);
+
+    /* The return value will be 0 when the peer has performed an orderly shutdown. */
+    if (*bytes_read == 0) {
+        *bytes_read = 0;
+        s2n_atomic_flag_set(&conn->read_closed);
+        RESULT_BAIL(S2N_ERR_CLOSED);
+    } else if (*bytes_read < 0) {
+        if (errno == EWOULDBLOCK || errno == EAGAIN) {
+            RESULT_BAIL(S2N_ERR_IO_BLOCKED);
+        }
+        RESULT_BAIL(S2N_ERR_IO);
+    }
+    *blocked = S2N_NOT_BLOCKED;
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_parse_ancillary_data(struct msghdr *msg, uint8_t *record_type)
+{
+    RESULT_ENSURE_REF(msg);
+    RESULT_ENSURE_REF(record_type);
+
+#if !S2N_KTLS_SUPPORTED /* CMSG_* macros are platform specific */
+    RESULT_BAIL(S2N_ERR_KTLS_UNSUPPORTED_PLATFORM);
+#else
+    RESULT_ENSURE_REF(msg->msg_control);
+    RESULT_ENSURE_GTE(msg->msg_controllen, CMSG_SPACE(sizeof(uint8_t)));
+
+    /* attempt to read the ancillary data */
+    struct cmsghdr *hdr = CMSG_FIRSTHDR(msg);
+    RESULT_ENSURE(hdr != NULL, S2N_ERR_IO);
+
+    /* iterate over all headers until it matches RECORD_TYPE.
+     * CMSG_NXTHDR will return NULL if there are no more cmsg */
+    while (hdr != NULL) {
+        if (hdr->cmsg_level == S2N_SOL_TLS && hdr->cmsg_type == S2N_TLS_GET_RECORD_TYPE) {
+            *record_type = *(unsigned char *) CMSG_DATA(hdr);
+            return S2N_RESULT_OK;
+        }
+
+        /* attempt to get the next header */
+        hdr = CMSG_NXTHDR(msg, hdr);
+    }
+#endif
+
+    /* return an IO error if no record was received */
+    RESULT_BAIL(S2N_ERR_IO);
+}
+
+/* Best practices taken from
+ * https://man7.org/tlpi/code/online/dist/sockets/scm_cred_recv.c.html */
+S2N_RESULT s2n_ktls_recv_msg(struct s2n_connection *conn, int sock, uint8_t *buf, size_t length,
+        uint8_t *record_type, s2n_blocked_status *blocked, ssize_t *bytes_read)
+{
+    RESULT_ENSURE_REF(buf);
+    RESULT_ENSURE_REF(record_type);
+    RESULT_ENSURE_REF(blocked);
+    RESULT_ENSURE_REF(bytes_read);
+    RESULT_ENSURE_GT(length, 0);
+
+    /* Init msghdr */
+    struct msghdr msg = { 0 };
+
+#if S2N_KTLS_SUPPORTED /* CMSG_* macros are platform specific */
```

Comment:
interleaving this check everywhere makes it a bit hard to follow, IMO. why not just wrap the whole file in this? none of these should even be called if kTLS isn't supported, right?

Reply:
could do that but thought we try to limit the `ifdef` usage to the minimum surface area. Thoughts?

Reply:
i would much rather read a single `ifdef` at the top of the file and say "ok this whole module is disabled if ktls is disabled" rather than having to make sure each individual function does the right thing.

Reply:
Love this idea. We don't so much try to limit ifdef surface area as keep ifdefs readable and minimal. If we can ifdef just one part of one method in a file, then great, keep the surface area as low as possible. But when more ifdefs start creeping in that becomes a problem and ifdefing the whole file definitely becomes the better option :)

But don't forget to have versions of any methods required outside of this file in the else. Otherwise the ifdefs just spread to other files.

Reply:
addressed in https://github.com/aws/s2n-tls/pull/4073/commits/0dfbfb3ad6976851327e5dd12019160799486ac6